### ISSUE TITLE
Reproduction Issue of ResNet-50 with dim=2048

### ISSUE THREAD

**Issue by chenshunpeng** [gmberton/EigenPlaces#4]
Hello, I thoroughly appreciated your series of work, showcasing a profound understanding of the VPR task. However, I've encountered challenges while attempting to reproduce the training on **ResNet-50 with dim=2048**, using a single 3090 GPU. I would greatly appreciate your assistance in identifying the issues and gaining a better grasp of the paper. For reference, I've provided the `info.log` file. Your help would be immensely valuable:

Mine:
< test - #q: 1000; #db: 2805840 >: R@1: 73.2, R@5: 79.7, R@10: 82.0, R@20: 84.0

Yours:
R@1: 84.1

Training:

```
2023-11-06 17:30:53   train.py --save_dir processed_Train
2023-11-06 17:30:53   Arguments: Namespace(M=15, N=3, focal_dist=10, s=100, m=0.4, lambda_lat=1.0, lambda_front=1.0, groups_num=9, min_images_per_class=5, backbone='ResNet50', fc_output_dim=2048, batch_size=32, epochs_num=40, iterations_per_epoch=5000, lr=1e-05, classifiers_lr=0.01, brightness=0.7, contrast=0.7, hue=0.5, saturation=0.7, random_resized_crop=0.5, infer_batch_size=16, positive_dist_threshold=25, resume_train=None, resume_model=None, device='cuda', seed=0, num_workers=8, visualize_classes=0, train_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/train', val_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/val', test_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/test', save_dir='processed_Train')
2023-11-06 17:30:53   The outputs are being saved in logs/processed_Train/2023-11-06_17-30-53
2023-11-06 17:30:53   Loading pretrained backbone's weights from CosPlace
2023-11-06 17:30:54   There are 1 GPUs and 64 CPUs.
2023-11-06 17:30:54   Using cached dataset cache/sfxl_M15_N3_mipc5.torch
2023-11-06 17:30:57   Using cached dataset cache/sfxl_M15_N3_mipc5.torch
2023-11-06 17:31:55   Using 18 groups
2023-11-06 17:31:55   The 18 groups have respectively the following number of classes {[len(g) for g in groups]}
2023-11-06 17:31:55   The 18 groups have respectively the following number of images {[g.get_images_num() for g in groups]}
2023-11-06 17:31:55   There are 5559 classes for the first group, each epoch has 5000 iterations with batch_size 32, therefore the model sees each class (on average) 28.8 times per epoch
2023-11-06 17:31:55   Validation set: < val - #q: 7983; #db: 8015 >
2023-11-06 17:31:55   Start training ...
2023-11-06 18:00:40   Epoch 00 in 0:28:45, < val - #q: 7983; #db: 8015 >: R@1: 90.4, R@5: 95.0, R@10: 96.1, R@20: 97.1
2023-11-06 18:29:30   Epoch 01 in 0:28:48, < val - #q: 7983; #db: 8015 >: R@1: 92.6, R@5: 96.1, R@10: 97.0, R@20: 97.8
2023-11-06 18:58:24   Epoch 02 in 0:28:52, < val - #q: 7983; #db: 8015 >: R@1: 92.6, R@5: 96.0, R@10: 97.0, R@20: 97.7
2023-11-06 19:27:21   Epoch 03 in 0:28:54, < val - #q: 7983; #db: 8015 >: R@1: 93.1, R@5: 96.4, R@10: 97.2, R@20: 98.0
2023-11-06 19:56:34   Epoch 04 in 0:29:11, < val - #q: 7983; #db: 8015 >: R@1: 92.8, R@5: 95.9, R@10: 96.8, R@20: 97.5
2023-11-06 20:26:30   Epoch 05 in 0:29:53, < val - #q: 7983; #db: 8015 >: R@1: 92.5, R@5: 95.7, R@10: 96.6, R@20: 97.4
2023-11-06 20:55:58   Epoch 06 in 0:29:25, < val - #q: 7983; #db: 8015 >: R@1: 92.4, R@5: 95.5, R@10: 96.6, R@20: 97.3
2023-11-06 21:26:05   Epoch 07 in 0:30:05, < val - #q: 7983; #db: 8015 >: R@1: 92.8, R@5: 96.2, R@10: 97.0, R@20: 97.8
2023-11-06 21:56:17   Epoch 08 in 0:30:10, < val - #q: 7983; #db: 8015 >: R@1: 93.2, R@5: 96.1, R@10: 97.0, R@20: 97.7
2023-11-06 22:26:18   Epoch 09 in 0:29:57, < val - #q: 7983; #db: 8015 >: R@1: 93.1, R@5: 96.0, R@10: 97.0, R@20: 97.6
2023-11-06 22:56:23   Epoch 10 in 0:30:01, < val - #q: 7983; #db: 8015 >: R@1: 93.9, R@5: 96.8, R@10: 97.6, R@20: 98.1
2023-11-06 23:26:21   Epoch 11 in 0:29:54, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.8, R@10: 97.6, R@20: 98.2
2023-11-06 23:56:17   Epoch 12 in 0:29:54, < val - #q: 7983; #db: 8015 >: R@1: 93.5, R@5: 96.6, R@10: 97.3, R@20: 97.8
2023-11-07 00:26:10   Epoch 13 in 0:29:49, < val - #q: 7983; #db: 8015 >: R@1: 94.1, R@5: 96.8, R@10: 97.6, R@20: 98.1
2023-11-07 00:56:25   Epoch 14 in 0:30:13, < val - #q: 7983; #db: 8015 >: R@1: 93.2, R@5: 96.2, R@10: 96.9, R@20: 97.6
2023-11-07 01:26:18   Epoch 15 in 0:29:49, < val - #q: 7983; #db: 8015 >: R@1: 92.9, R@5: 95.8, R@10: 96.5, R@20: 97.3
2023-11-07 01:56:30   Epoch 16 in 0:30:08, < val - #q: 7983; #db: 8015 >: R@1: 93.6, R@5: 96.4, R@10: 97.1, R@20: 97.8
2023-11-07 02:26:28   Epoch 17 in 0:29:54, < val - #q: 7983; #db: 8015 >: R@1: 93.9, R@5: 96.7, R@10: 97.5, R@20: 98.0
2023-11-07 02:56:37   Epoch 18 in 0:30:06, < val - #q: 7983; #db: 8015 >: R@1: 94.4, R@5: 96.9, R@10: 97.5, R@20: 98.0
2023-11-07 03:26:57   Epoch 19 in 0:30:16, < val - #q: 7983; #db: 8015 >: R@1: 93.7, R@5: 96.5, R@10: 97.2, R@20: 97.9
2023-11-07 03:57:18   Epoch 20 in 0:30:18, < val - #q: 7983; #db: 8015 >: R@1: 93.8, R@5: 96.5, R@10: 97.3, R@20: 97.9
2023-11-07 04:27:42   Epoch 21 in 0:30:20, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.8, R@10: 97.4, R@20: 97.9
2023-11-07 04:58:01   Epoch 22 in 0:30:15, < val - #q: 7983; #db: 8015 >: R@1: 94.2, R@5: 96.8, R@10: 97.3, R@20: 98.0
2023-11-07 05:28:23   Epoch 23 in 0:30:19, < val - #q: 7983; #db: 8015 >: R@1: 93.7, R@5: 96.6, R@10: 97.2, R@20: 97.8
2023-11-07 05:58:43   Epoch 24 in 0:30:17, < val - #q: 7983; #db: 8015 >: R@1: 93.7, R@5: 96.4, R@10: 97.2, R@20: 97.7
2023-11-07 06:29:06   Epoch 25 in 0:30:19, < val - #q: 7983; #db: 8015 >: R@1: 94.1, R@5: 96.8, R@10: 97.5, R@20: 98.0
2023-11-07 06:59:31   Epoch 26 in 0:30:22, < val - #q: 7983; #db: 8015 >: R@1: 93.8, R@5: 96.6, R@10: 97.2, R@20: 97.8
2023-11-07 07:29:53   Epoch 27 in 0:30:18, < val - #q: 7983; #db: 8015 >: R@1: 93.7, R@5: 96.3, R@10: 97.1, R@20: 97.6
2023-11-07 08:00:18   Epoch 28 in 0:30:22, < val - #q: 7983; #db: 8015 >: R@1: 94.3, R@5: 97.1, R@10: 97.6, R@20: 98.2
2023-11-07 08:30:42   Epoch 29 in 0:30:20, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.7, R@10: 97.3, R@20: 97.9
2023-11-07 09:01:24   Epoch 30 in 0:30:39, < val - #q: 7983; #db: 8015 >: R@1: 93.5, R@5: 96.4, R@10: 97.0, R@20: 97.6
2023-11-07 09:31:50   Epoch 31 in 0:30:20, < val - #q: 7983; #db: 8015 >: R@1: 93.9, R@5: 96.6, R@10: 97.3, R@20: 97.9
2023-11-07 10:02:30   Epoch 32 in 0:30:33, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.7, R@10: 97.4, R@20: 98.0
2023-11-07 10:32:55   Epoch 33 in 0:30:22, < val - #q: 7983; #db: 8015 >: R@1: 93.7, R@5: 96.4, R@10: 97.2, R@20: 97.8
2023-11-07 11:04:10   Epoch 34 in 0:31:12, < val - #q: 7983; #db: 8015 >: R@1: 93.8, R@5: 96.5, R@10: 97.2, R@20: 97.8
2023-11-07 11:35:41   Epoch 35 in 0:31:27, < val - #q: 7983; #db: 8015 >: R@1: 94.1, R@5: 96.7, R@10: 97.4, R@20: 97.8
2023-11-07 12:07:15   Epoch 36 in 0:31:31, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.6, R@10: 97.3, R@20: 97.8
2023-11-07 12:38:48   Epoch 37 in 0:31:30, < val - #q: 7983; #db: 8015 >: R@1: 94.0, R@5: 96.8, R@10: 97.5, R@20: 98.0
2023-11-07 13:10:13   Epoch 38 in 0:31:21, < val - #q: 7983; #db: 8015 >: R@1: 93.8, R@5: 96.6, R@10: 97.1, R@20: 97.8
2023-11-07 13:41:33   Epoch 39 in 0:31:17, < val - #q: 7983; #db: 8015 >: R@1: 93.9, R@5: 96.5, R@10: 97.3, R@20: 97.8
2023-11-07 13:41:36   Trained for 40 epochs, in total in 20:10:43
```

Eval:
```
2023-11-10 17:30:05   eval.py --backbone ResNet50 --fc_output_dim 2048 --save_dir processed_test_train --resume_model /work/chenshunpeng/project/VPR/EigenPlaces-main/EigenPlaces-main/logs/processed_Train/2023-11-06_17-30-53/best_model.pth
2023-11-10 17:30:05   Arguments: Namespace(M=15, N=3, focal_dist=10, s=100, m=0.4, lambda_lat=1.0, lambda_front=1.0, groups_num=9, min_images_per_class=5, backbone='ResNet50', fc_output_dim=2048, batch_size=32, epochs_num=40, iterations_per_epoch=5000, lr=1e-05, classifiers_lr=0.01, brightness=0.7, contrast=0.7, hue=0.5, saturation=0.7, random_resized_crop=0.5, infer_batch_size=16, positive_dist_threshold=25, resume_train=None, resume_model='/work/chenshunpeng/project/VPR/EigenPlaces-main/EigenPlaces-main/logs/processed_Train/2023-11-06_17-30-53/best_model.pth', device='cuda', seed=0, num_workers=8, visualize_classes=0, train_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/train', val_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/val', test_dataset_folder='/work/chenshunpeng/data/sf_xl/processed/test', save_dir='processed_test_train')
2023-11-10 17:30:05   The outputs are being saved in logs/processed_test_train/2023-11-10_17-30-04
2023-11-10 17:30:05   Loading pretrained backbone's weights from CosPlace
2023-11-10 17:30:08   There are 1 GPUs and 64 CPUs.
2023-11-10 17:30:08   Loading model_ from /work/chenshunpeng/project/VPR/EigenPlaces-main/EigenPlaces-main/logs/processed_Train/2023-11-06_17-30-53/best_model.pth
2023-11-10 22:43:39   < test - #q: 1000; #db: 2805840 >: R@1: 73.2, R@5: 79.7, R@10: 82.0, R@20: 84.0
```

Besides, I get the same results as in the paper using the released trained weights.
`test/queries_v1`:
< test - #q: 1000; #db: 2805840 >: R@1: 84.1, R@5: 89.1, R@10: 90.8, R@20: 92.6

Based on the large dataset size, I'd like to verify if there were any issues during the dataset download process. Below are the sizes, in MB, of the subfiles within the processed dataset, specifically in the training and validation sections:

`sf_xl/processed/train`:
MB	Folder
10008   37.70
14002   37.71
13916   37.72
22513   37.73
20563   37.74
24293   37.75
32104   37.76
35248   37.77
35154   37.78
25809   37.79
13383   37.80
34      37.81

`sf_xl/processed/val`:
MB	Folder
357     database
355     queries

Thank you for your help!
@gmberton

**Comment by gmberton** [comment]
Hi @chenshunpeng, thanks a lot!
The hyperparameters look correct. I just realized that there is a small logging bug at [this line of the train.py](https://github.com/gmberton/EigenPlaces/blob/main/train.py#L72), because the info.log should show the number of training images. That does not change the training, but it makes the log more clear.
Could you show me the number of images within `sf_xl/processed/train`? The size in MB is not really relevant as it is dependent on the file system

**Comment by chenshunpeng** [comment]
Of course, here's the breakdown of the image count in each folder:
```
(EigenPlaces) chenshunpeng@h3c-R5300-G5:~/data/sf_xl/processed/train$ find . -maxdepth 1 -mindepth 1 -type d -exec sh -c 'echo -n "{}: "; find "{}" -maxdepth 1 -type f \( -iname \*.jpg -o -iname \*.jpeg -o -iname \*.png \) | wc -l' \;
./37.70: 256914
./37.71: 343182
./37.72: 347208
./37.73: 538302
./37.74: 482880
./37.75: 541038
./37.76: 714156
./37.77: 785682
./37.78: 753486
./37.79: 535620
./37.80: 307878
./37.81: 1074
```

**Comment by gmberton** [comment]
Ok so the issue seems to be in the number of training images, because I have this amount of images
```
37.70 163665
37.71 228643
37.72 235935
37.73 347857
37.74 311148
37.75 335999
37.76 412361
37.77 458764
37.78 438066
37.79 309385
37.80 188416
37.81 853
```
It's quite strange that you have less than me. Perhaps instead of downloading the `panoramas` folder you downloaded something else? You can download the `all_images_paths.txt` file and see which images are mismatched

**Comment by chenshunpeng** [comment]
Okay, I understand. My training set is incomplete. I will re-download the whole 360° panorama and re-enact it. Thank you for your guidance.
